AzureDevOpsPipelines/deployment/bot-publish/source/Core.ts

```diff
@@ -0,0 +1,110 @@
+import tl = require("azure-pipelines-task-lib");
+import { IExecSyncResult, IExecOptions } from 'azure-pipelines-task-lib/toolrunner';
+import { execSync } from "child_process";
+import path = require("path");
+import fs = require("fs");
+
+export class core {
+    private isLoggedIn: boolean = false;
+    private cliPasswordPath: string = "";
+    private servicePrincipalId: string = "";
+    private servicePrincipalKey: string = "";
+    private prefix: string = "";
+    
+    public subscriptionID: string = "";
+
+    constructor() {
+        this.getPrefix();
+        this.subscriptionID = tl.getVariable('subscriptionID');
+        
+        if (this.subscriptionID == null) {
+            this.loginAzureRM(tl.getInput("connectedServiceNameARM", true));
+            tl.setVariable('subscriptionID', this.subscriptionID);
+        }
+    }
+
+    private loginAzureRM(connectedService: string): void {
+        var authScheme: string = tl.getEndpointAuthorizationScheme(connectedService, true);
+        this.subscriptionID = tl.getEndpointDataParameter(connectedService, "SubscriptionID", true);
+        if(authScheme.toLowerCase() == "serviceprincipal") {
+            let authType: string = tl.getEndpointAuthorizationParameter(connectedService, 'authenticationType', true);
+            let cliPassword: string;
+            var servicePrincipalId: string = tl.getEndpointAuthorizationParameter(connectedService, "serviceprincipalid", false);
+            if (authType == "spnCertificate") {
+                tl.debug('certificate based endpoint');
+                let certificateContent: string = tl.getEndpointAuthorizationParameter(connectedService, "servicePrincipalCertificate", false);
+                cliPassword = path.join(tl.getVariable('Agent.TempDirectory') || tl.getVariable('system.DefaultWorkingDirectory'), 'spnCert.pem');
+                fs.writeFileSync(cliPassword, certificateContent);
+                this.cliPasswordPath = cliPassword;
+
+            }
+            else {
+                tl.debug('key based endpoint');
+                cliPassword = tl.getEndpointAuthorizationParameter(connectedService, "serviceprincipalkey", false);
+                this.servicePrincipalId = servicePrincipalId;
+                this.servicePrincipalKey = cliPassword;
+            }
+            var tenantId: string = tl.getEndpointAuthorizationParameter(connectedService, "tenantid", false);
```

Comment:
If it's value is only set once here, it'd be a const instead of var. Also, can we put it on top with all the variables declarations? If not, add a space between line 46-47

Reply:
Can we review this?